### PR TITLE
V2 OctoPrint Websocket and API Client

### PR DESCRIPTION
**Pull request by davidzwa**
- [x] Adds websocket package `ws`
- [x] Experiment with websockets
- [x] Adds client API service 
- [x] Expresses scaffold for client state management
- [x] Test JSON to Typescript schema
- [x] Test Global API Key detection
- [x] Test connection state re-initiation protection
- [x] Test connection sealed/frozen

Printer view now has:
- [x] No null exceptions when user group is null (admin buttons)
- [x] Add printer possibility (sortIndex now auto-generated)

Still need to work on the user, but I'm too lazy to enable passport cookies/sesh.

Oh and fun feature: can add a test printer to check at start through `TEST_PRINTER_URL` `TEST_PRINTER_KEY`!

### CODE REVIEW

File: server_2x/src/octoprint/gateway/octoprint.gateway.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import {WebSocketGateway} from '@nestjs/websockets';
+import {OctoPrintClientService} from '../services/octoprint-client.service';
+
+@WebSocketGateway()
+export class OctoprintGateway {
+    constructor(
+        private readonly octoprintService: OctoPrintClientService
+    ) {
+        this.connectOctoPrintClient();
```
Comment (davidzwa) [inline server_2x/src/octoprint/gateway/octoprint.gateway.ts:9]: Should remove this example code. Especially in a constructor!

File: server_2x/src/octoprint/gateway/octoprint.gateway.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    connectOctoPrintClient() {
+        // const socket = new WebSocket('ws://prusa');
```
Comment (davidzwa) [inline server_2x/src/octoprint/gateway/octoprint.gateway.ts:13]: Remove...

File: server_2x/src/octoprint/octoprint.module.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        const testPrinterConnectionParams = this.getEnvTestPrinter();
+        this.service.getSettings(testPrinterConnectionParams)
+            .subscribe(result => {
+                testPrinterConnectionParams.printerKey
+                console.warn("Global key provided", result.data.api.key === testPrinterConnectionParams.printerKey);
+            });
+    }
+
+    private getEnvTestPrinter(): ConnectionParamsModel {
+        const envTestPrinterURL = process.env["TEST_PRINTER_URL"];
```
Comment (davidzwa) [inline server_2x/src/octoprint/octoprint.module.ts:34]: Constants instead? 
And maybe parse a YAML config to allow multiple printers at once?

File: server_2x/src/octoprint/services/octoprint-client.service.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            throw Error("Can't connect to printer without printer API Key");
+        }
+    }
+
+    getSettings(connectionParams: ConnectionParamsModel): Observable<AxiosResponse<OctoPrintSettingsDto>> {
+        this.checkConnectionParams(connectionParams);
+
+        return this.httpService.get<OctoPrintSettingsDto>(new URL('api/settings', connectionParams.printerURL).toString(), {
+            headers: {
+                "x-api-key": connectionParams.printerKey
```
Comment (davidzwa) [inline server_2x/src/octoprint/services/octoprint-client.service.ts:28]: Hmm maybe use an interceptors, but avoid repeating this...

File: server_2x/src/octoprint/services/octoprint-client.service.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        this.checkConnectionParams(params);
+
+        return this.httpService.post<OctoPrintSettingsDto>(new URL('api/settings', params.printerURL).toString(),
+            {
+                "api": {
+                    "allowCrossOrigin": true
+                }
+            }, {
+                headers: [{
+                    "x-api-key": params.printerKey
```
Comment (davidzwa) [inline server_2x/src/octoprint/services/octoprint-client.service.ts:43]: Same problem about interception.

File: server_2x/src/octoprint/services/web-sockets-client.service.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import {Injectable} from "@nestjs/common";
+import * as WebSocket from "ws";
+
+@Injectable()
+export class WebSocketsClientService {
+    // https://stackoverflow.com/questions/61652247/connect-nestjs-to-a-websocket-server
+    // wss://echo.websocket.org is a test websocket server
+    private ws = new WebSocket("wss://echo.websocket.org");
```
Comment (davidzwa) [inline server_2x/src/octoprint/services/web-sockets-client.service.ts:8]: Clear out test code

File: server_2x/src/octoprint/state/client-connection.state.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    setCorsEnabled(params: ConnectionParamsModel): Observable<any> {
+        if (this.userHasRequiredRoles()) {
+            return this.octoPrintClientService
+                .setCORSEnabled(params)
+                .pipe(tap(_ => {
+                    this.patchState({
+                        corsEnabled: true
+                    });
+                }, error => {
+                    // Analyze the state
```
Comment (davidzwa) [inline server_2x/src/octoprint/state/client-connection.state.ts:63]: Didnt test?

File: server_2x/src/printers/tasks/printer-query.task.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    //     this.logger.warn('Called every 55 milli-seconds');
+    // }
+    //
+    // @Interval(233)
+    // handleIntervalFast() {
+    //     this.logger.warn('Called every 233 milli-seconds');
+    // }
+
+    @Timeout(5000)
+    handleTimeout() {
```
Comment (davidzwa) [inline server_2x/src/printers/tasks/printer-query.task.ts:26]: Remove test-code.

File: tools/client-proxy/api.analyzer.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+// import * as fs from "fs";
+// import {transform} from 'json-to-typescript';
```
Comment (davidzwa) [inline tools/client-proxy/api.analyzer.ts:2]: Add comments as to why this is here and what the future plans of this tool are... useless like this.

**Review by NotExpectedYet — APPROVED** [review]
This i love! 

Check's and all, makes sex on wheels.

**Pull request merged**